packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -3,30 +3,489 @@
+    }
+
+    /**
+     * @notice close short position with user Uniswap v3 LP NFT
+     * @dev user should approve this contract for Uni NFT transfer
+     */
+    function closeShortWithUserNft(CloseShortWithUserNftParams calldata params) external {
+        INonfungiblePositionManager(nonfungiblePositionManager).safeTransferFrom(
+            msg.sender,
+            address(this),
+            params.tokenId
+        );
+
+        // get liquidity amount, and withdraw ETH and wPowerPerp amounts in LP position
+        (uint128 liquidity, , ) = ControllerHelperLib._getUniPositionBalances(
+            nonfungiblePositionManager,
+            params.tokenId,
+            IOracle(oracle).getTimeWeightedAverageTickSafe(wPowerPerpPool, 420),
+            isWethToken0
+        );
+        // (
+        //     ,
+        //     ,
+        //     ,
+        //     ,
+        //     ,
+        //     ,
+        //     ,
+        //     uint128 liquidity,
+        //     ,
+        //     ,
+        //     ,
+        // ) = INonfungiblePositionManager(nonfungiblePositionManager).positions(params.tokenId);
+        INonfungiblePositionManager.DecreaseLiquidityParams memory decreaseParams = INonfungiblePositionManager
+            .DecreaseLiquidityParams({
+                tokenId: params.tokenId,
+                liquidity: uint128(uint256(liquidity).mul(params.liquidityPercentage).div(1e18)),
+                amount0Min: params.amount0Min,
+                amount1Min: params.amount1Min,
+                deadline: block.timestamp
+            });
+
+        INonfungiblePositionManager(nonfungiblePositionManager).decreaseLiquidity(decreaseParams);
+
+        uint256 wethAmount;
+        uint256 wPowerPerpAmount;
+        (isWethToken0)
+            ? (wethAmount, wPowerPerpAmount) = INonfungiblePositionManager(nonfungiblePositionManager).collect(
+                INonfungiblePositionManager.CollectParams({
+                    tokenId: params.tokenId,
+                    recipient: address(this),
+                    amount0Max: type(uint128).max,
+                    amount1Max: type(uint128).max
+                })
+            )
+            : (wPowerPerpAmount, wethAmount) = INonfungiblePositionManager(nonfungiblePositionManager).collect(
+            INonfungiblePositionManager.CollectParams({
+                tokenId: params.tokenId,
+                recipient: address(this),
+                amount0Max: type(uint128).max,
+                amount1Max: type(uint128).max
+            })
+        );
+
+        if (wPowerPerpAmount < params.wPowerPerpAmountToBurn) {
+            // swap needed wPowerPerp amount to close short position
+            // TODO: need to set max slippage here
+            _exactOutFlashSwap(
+                weth,
+                wPowerPerp,
+                IUniswapV3Pool(wPowerPerpPool).fee(),
+                params.wPowerPerpAmountToBurn.sub(wPowerPerpAmount),
+                IWETH9(weth).balanceOf(address(this)),
```

Comment:
need to be able to set this slippage as well

this could be a very large number if a lot of ETH is withdrawn against a small short position (so effectively have something like 1000 WETH to buy 1 oSQTH)

Reply:
2 ways I can think about it 
- f/e estimates if user getting excess or shortfall and sets the limit price based on that, this will fail if the direction switches as the limit will be breached
- f/e sets a buy and sell px (probably overkill for the above edge case?) 

We have to set slippage for a user closing a short with a LP position.  If the LP oSQTH is < shortAmount, they buy more (and need a limit price).  If the LP oSQTH is > shortAmount they need to sell oSQTH and need a limit price.

@KMKoushik @alexisgauba , any thoughts?

Reply:
can rename the slippage param to something that doesnt specify a direction if go that way - ie limitPriceEthPerPowerPerp